shopinvader_validation_token/models/shopinvader_security_token.py

```diff
@@ -0,0 +1,167 @@
+# -*- coding: utf-8 -*-
+# Copyright 2021 ACSONE SA/NV (<http://acsone.eu>)
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html).
+from random import randint
+
+from dateutil.relativedelta import relativedelta
+
+from odoo import api, fields, models
+
+from odoo.addons.base.res.res_partner import _lang_get
+
+
+class ShopinvaderSecurityToken(models.Model):
+    _name = "shopinvader.security.token"
+    _description = "Shopinvader security token"
+
+    email = fields.Char(required=True,)
+    expiration_datetime = fields.Datetime(
+        string="Token's expiration", required=True, readonly=True,
+    )
+    token = fields.Char(required=True, index=True, readonly=True,)
+    active = fields.Boolean(
+        default=True, help="Become inactive when the token is used.",
+    )
+    shopinvader_backend_id = fields.Many2one(
+        comodel_name="shopinvader.backend", required=True,
+    )
+    notification_type = fields.Char(required=True)
+    lang = fields.Selection(
+        _lang_get, string="Language", default=lambda self: self.env.lang,
+    )
+
+    _sql_constraints = [
+        ("uniq_token", "unique(token)", "This token already exists!"),
+    ]
+
+    @api.multi
+    def _consume(self):
+        """
+
+        :return: bool
+        """
+        return self.sudo().write({"active": False})
+
+    @api.model
+    def _disable_expired_token(self):
+        """Disable every expired or un-used token.
+
+        Should be used in a cron.
+        :return:
+        """
+        now = fields.Datetime.now()
+        domain = [
+            "|",
+            ("active", "=", False),
+            ("expiration_datetime", "<", now),
+        ]
+        self.sudo().search(domain).unlink()
+
+    @api.model
+    def _service_notification_map(self):
+        """
+
+        :return: dict
+        """
+        return {
+            "shopinvader.customer.service": "partner_token_customer",
+        }
+
+    @api.model
+    def _should_trigger_security_token(self, backend, email, service_name):
+        """Check if the notification partner_token_* is enabled on the related backend.
+
+        As we have the backend, email and the service who do this check,
+        we can precisely check if the token is required or not.
+        :param backend: shopinvader.backend recordset
+        :param email: str
+        :param service_name: str
+        :return: bool
+        """
+        # We have to force the backend when the current shopinvader.partner
+        # doesn't exist yet.
+        backend.ensure_one()
+        should_trigger = False
+        if service_name and backend:
+            notif_type = self._service_notification_map().get(service_name, "")
+            if notif_type:
+                should_trigger = any(
+                    n.notification_type == notif_type
```

Comment:
I don't get the reason of this logic: the token stuff is enabled only if you have the notification? :thinking: